pkg/extract/oracle/oracle.go

```diff
@@ -41,24 +72,482 @@ func Extract(args string, opts ...Option) error {
+				Criteria: criteriaTypes.Criteria{
+					Operator: criteriaTypes.CriteriaOperatorTypeOR,
+					Criterions: func() []criterionTypes.Criterion {
+						criterions := make([]criterionTypes.Criterion, 0, len(pkgs))
+						for pkg, archs := range pkgs {
+							criterions = append(criterions, criterionTypes.Criterion{
+								Vulnerable: true,
+								Package: packageTypes.Package{
+									Name:          pkg.name,
+									Architectures: archs,
+								},
+								Affected: &affectedTypes.Affected{
+									Type:  rangeTypes.RangeTypeRPM,
+									Range: []rangeTypes.Range{{LessThan: pkg.fixedVersion}},
+									Fixed: []string{pkg.fixedVersion},
+								},
+							})
+						}
+						return criterions
+					}(),
+				},
+			})
+		}
+		return ds
+	}()
+
+	return data, nil
+}
+
+type ovalPackage struct {
+	major           string
+	ovalName        string
+	fixedVersion    string
+	modularityLabel string
+	arch            string
+}
+
+func collectPackages(criteria oracle.Criteria, tos tos) (packages, error) {
+	pkgs, err := evalCriteria(criteria, tos)
+	if err != nil {
+		return packages{}, errors.Wrapf(err, "applyCriteria")
+	}
+
+	allPkgs := packages{}
+	for _, p := range pkgs {
+		pkgs, ok := allPkgs[majorType(p.major)]
+		if !ok {
+			pkgs = perMajorPackages{}
+			allPkgs[majorType(p.major)] = pkgs
+		}
+		key := packageKey{
+			name: func() string {
+				switch p.modularityLabel {
+				case "":
+					return p.ovalName
+				default:
+					return fmt.Sprintf("%s::%s", p.modularityLabel, p.ovalName)
+				}
+			}(),
+			fixedVersion: p.fixedVersion,
+		}
+		pkgs[key] = append(pkgs[key], p.arch)
+	}
+	return allPkgs, nil
+}
+
+func evalCriteria(criteria oracle.Criteria, tos tos) ([]ovalPackage, error) {
+	// Exclude patterns that do not exists in oracle oval data and don't implement them, YAGNI.
+	// With these constraints, we can ignore Criteria.Operator (AND or OR) to extract package information.
+	switch {
+	case criteria.Operator == "OR" && len(criteria.Criterions) > 0:
+		return []ovalPackage{}, errors.Errorf("criteriaons under OR criteria MUST not exist")
+	case criteria.Operator == "OR" && len(criteria.Criterias) == 0:
+		return []ovalPackage{}, errors.Errorf("criterias under OR criteria MUST exist")
+	case criteria.Operator == "AND" && len(criteria.Criterias) > 1:
+		return []ovalPackage{}, errors.Errorf("criterias under AND-criteria MUST be < 2")
+	}
+
+	var pkgs []ovalPackage
+
+	for _, ca := range criteria.Criterias {
+		ps, err := evalCriteria(ca, tos)
+		if err != nil {
+			return []ovalPackage{}, err
+		}
+		pkgs = append(pkgs, ps...)
+	}
+
+	if err := evalCriterions(&pkgs, tos, criteria.Criterions); err != nil {
+		return []ovalPackage{}, err
+	}
+	return pkgs, nil
+}
+
+func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion) error {
+	if len(criterions) == 0 {
+		return nil
+	}
+
+	// The case for parent criterion is AND and without child criterions case. Add base (ANY) one
```

Comment:
typo?
```suggestion
	// The case for parent criteria is AND and without child criterions case. Add base (ANY) one
```

Reply:
If you need to add elements to pkgs based on criteria, shouldn't you do that in `evalCriteria`?
```diff
:100644 100644 f41fe0ef 00000000 M	pkg/extract/oracle/oracle.go

diff --git a/pkg/extract/oracle/oracle.go b/pkg/extract/oracle/oracle.go
index f41fe0ef..66fa8f93 100644
--- a/pkg/extract/oracle/oracle.go
+++ b/pkg/extract/oracle/oracle.go
@@ -325,22 +325,22 @@ func evalCriteria(criteria oracle.Criteria, tos tos) ([]ovalPackage, error) {
 		pkgs = append(pkgs, ps...)
 	}
 
-	if err := evalCriterions(&pkgs, tos, criteria.Criterions); err != nil {
+	// // The case for parent criteria is AND and without child criterias case. Add base (ANY) one
+	if criteria.Operator == "AND" && len(criteria.Criterias) == 0 {
+		pkgs = append(pkgs, ovalPackage{})
+	}
+
+	if err := evalCriterions(pkgs, tos, criteria.Criterions); err != nil {
 		return []ovalPackage{}, err
 	}
 	return pkgs, nil
 }
 
-func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion) error {
+func evalCriterions(pkgs []ovalPackage, tos tos, criterions []oracle.Criterion) error {
 	if len(criterions) == 0 {
 		return nil
 	}
 
-	// The case for parent criterion is AND and without child criterions case. Add base (ANY) one
-	if len(*pkgs) == 0 {
-		*pkgs = append(*pkgs, ovalPackage{})
-	}
-
 	for _, c := range criterions {
 		test, ok := tos.rpminfoTests[c.TestRef]
 		if ok {
@@ -357,12 +357,12 @@ func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion)
 			case "oraclelinux-release":
 				switch {
 				case state.Version != nil:
-					for i := range *pkgs {
-						(*pkgs)[i].major = strings.TrimPrefix(state.Version.Text, "^")
+					for i := range pkgs {
+						pkgs[i].major = strings.TrimPrefix(state.Version.Text, "^")
 					}
 				case state.Arch != nil:
-					for i := range *pkgs {
-						(*pkgs)[i].arch = state.Arch.Text
+					for i := range pkgs {
+						pkgs[i].arch = state.Arch.Text
 					}
 				default:
 					return errors.Errorf("invalid combination, obj %+v and state %+v", obj, state)
@@ -373,9 +373,9 @@ func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion)
 					if state.Evr.Operation != "less than" {
 						return errors.Errorf("unexpected evr operation: %s", state.Evr.Operation)
 					}
-					for i := range *pkgs {
-						(*pkgs)[i].ovalName = obj.Name
-						(*pkgs)[i].fixedVersion = state.Evr.Text
+					for i := range pkgs {
+						pkgs[i].ovalName = obj.Name
+						pkgs[i].fixedVersion = state.Evr.Text
 					}
 				case state.SignatureKeyid != nil:
 				case state.Release != nil:
@@ -410,8 +410,8 @@ func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion)
 			streamWithBackSlashes, _, _ := strings.Cut(strings.TrimPrefix(state.Text.Text, `\nstream\s*=\s*`), `\b[\w\W]`)
 			// There may be "." in stream and should be unescaped
 			stream := strings.ReplaceAll(streamWithBackSlashes, `\`, "")
-			for i := range *pkgs {
-				(*pkgs)[i].modularityLabel = fmt.Sprintf("%s:%s", module, stream)
+			for i := range pkgs {
+				pkgs[i].modularityLabel = fmt.Sprintf("%s:%s", module, stream)
 			}
 		}
 	}
```

Reply:
If we assume that Oracle's OVAL criteria has a structure starting from the root, which is major → arch → (modularity label) → package, we can also write it like this.
```diff
:100644 100644 f41fe0ef 00000000 M	pkg/extract/oracle/oracle.go

diff --git a/pkg/extract/oracle/oracle.go b/pkg/extract/oracle/oracle.go
index f41fe0ef..9455936c 100644
--- a/pkg/extract/oracle/oracle.go
+++ b/pkg/extract/oracle/oracle.go
@@ -336,11 +336,6 @@ func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion)
 		return nil
 	}
 
-	// The case for parent criterion is AND and without child criterions case. Add base (ANY) one
-	if len(*pkgs) == 0 {
-		*pkgs = append(*pkgs, ovalPackage{})
-	}
-
 	for _, c := range criterions {
 		test, ok := tos.rpminfoTests[c.TestRef]
 		if ok {
@@ -373,10 +368,10 @@ func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion)
 					if state.Evr.Operation != "less than" {
 						return errors.Errorf("unexpected evr operation: %s", state.Evr.Operation)
 					}
-					for i := range *pkgs {
-						(*pkgs)[i].ovalName = obj.Name
-						(*pkgs)[i].fixedVersion = state.Evr.Text
-					}
+					*pkgs = append(*pkgs, ovalPackage{
+						ovalName:     obj.Name,
+						fixedVersion: state.Evr.Text,
+					})
 				case state.SignatureKeyid != nil:
 				case state.Release != nil:
 					// <rpminfo_state xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#linux" id="oval:com.oracle.elsa:ste:20163515002" version="501">
```

Reply:
> I don't like this style of writing because I feel it goes against the basic principle of setting the contents of criterion for each element of pkgs in evalCriterions.

Reply:
> If you need to add elements to pkgs based on criteria, shouldn't you do that in evalCriteria?

Chose this! by 52d64ad